src/Lecture2.hs

```diff
@@ -282,7 +368,16 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval vars expr = case expr of
+    Lit x -> Right x
+    Var x -> case lookup x vars of
+        Just var -> Right var
+        Nothing -> Left (VariableNotFound x)
+    Add lhs rhs -> case (eval vars lhs, eval vars rhs) of
+        (Left (VariableNotFound a), _) -> Left (VariableNotFound a)
+        (_, Left (VariableNotFound b)) -> Left (VariableNotFound b)
+        (Right a, Right b) -> Right (a + b)
```

Comment:
That's a great case matching! 👏🏻 
Once you learn about Monads, you'll be able to reduce this function to a single line ✂️